core/src/main/scala/cats/TFunctor.scala

```diff
@@ -0,0 +1,42 @@
+package cats
+
+import simulacrum.typeclass
+
+/**
+ * This is an endofunctor in the category of endofunctors in `Skal`.
+ *
+ * `Skal` is the category of scala types. Functors in `Skal`
+ * is encoded as `Functor`. The functors in `Skal` themselves forms
+ * a category, let's denote it as `F[Skal]`.
+ * In `F[Skal]`, functors of Skal, e.g. `Option[_]` and `Either[E, _]`, are objects,
+ * while natural transformations, e.g. `Option ~> Either[E, ?]`, are arrows.
+ * A endofunctor in `F[Skal]` maps one set of functors of `Skal` to another
+ * set of functors of `Skal` while preserving the structures.
+ *
+ * For `TFunctor`, the domain is `F[_]`, the codomain is `T[F, _]`, both are
+ * functors in `Skal`. The `TFunctor` provides a mapping from the arrows between
+ * `F[_]` and `G[_]`, i.e. `F ~> G` to arrows between `T[F, _]` and `T[G, _]`,
+ * i.e. `T[F, ?] ~> T[G, ?]`. The `lift` method makes this intention clear.
+ *
+ * In `cats.core`, examples of such `TFunctor`s are monad transformers such
+ * as `OptionT`, `EitherT`
+ *
+ */
+@typeclass trait TFunctor[T[_[_], _]] {
```

Comment:
do we need `T[_[_], _]` or would `T[_[_]]` do? I'm not sure what the second parameter is buying us. Maybe I'm missing it. Can't we just ignore the inner value type? Also, wouldn't we have more flexibility only acting on `T[_[_]]`?

Reply:
The main reason I chose `T[_[_], _]` here, is that given any `Functor` `F[_]`,  `T[F, _]` is also a `Functor`. So `T[_[_], _]` can be deemed as a *functor* between the `F[_]` `Functor` and the `T[F, _]` `Functor`. i.e. a *endofunctor* in the category of *endofunctor*s, thus more like a higher kinded `Functor`. 
If you use `T[_[_]]`, then given any `Functor` `F[_]`, `T[F]` isn't a `Functor`, it's a kind * type, not a type constructor. So `T[_[_]]` is a *functor* between `F[_]` `Functor` and kind * type `T[F]`s. That makes it less a high kinded `Functor`. However, you can work around with the parametricity in instance definitions, e.g. for `OptionT`, 
```scala
def functorKInstance[A]: FunctorK[OptionT[?[_], A]]
```
With parametric `A`, we can have a `FunctorK` instance forall `A`s.  In some sense, we may choose to see this set of `FunctorK` instances as a single truely higher kinded *functor*.

Practical use wise, I can't think of any downside of using `T[_[_]]`. The only downside is its awkward encoding of the higher kinded `Functor` - it relies on instance definition. It may have more usecases, one that immediately jumps out is `FunctionK`. So maybe practically it's more useful. 

Now, the fact that we aren't sure about this thing makes me think that maybe we shouldn't place this in `cats-core`. I already have `FunctorK[T[_]]` defined in mainecoon, I could add all these instances there. Or we can create a new `cats-extra` module for this type of less commonly used type class.

Reply:
With the example of `FunctionK`, my first thought is that that’s what’s defined incorrectly. We really want `FunctionK[F[_], G[_], A]` and `Forall[FunctionK[F, G, ?]]` for a natural transformation. Then you _do_ have a `FunctorK[FunctionK[F, ?[_], ?]]`. But that might just be crazy talk.

Reply:
I'm not sure if "higher-kinded X" is a really rigid term. In my mind, higher-kinded functor could equally well mean a functor from (endofunctors in Scal) to Scal, or a functor from endofunctors in Scal to other endofunctors in Scal. This is my main problem with the `-K` naming scheme; it gets even worse when type classes have multiple type parameters.

As well, the encoding proposed above does not reify that the `A` can vary with the instance remaining valid. The consumer of the instance needs a `Forall[Lambda[A =>FunctorK[OptionT[?[_], A]]]]` to have that fact in hand, and that makes it much more unwieldy, though a type alias could help.

To my mind if we're going to do a cats-extras project we'll want this there.

Reply:
I agree with @kailuowang and @edmundnoble, I don't think it's all that useful in cats-core. I'd like to see conformity in Monad Transformers for defining a `mapK` method, but it doesn't need a type class in core IMO.

Reply:
> I'm not sure if "higher-kinded X" is a really rigid term.

Absolutely. I don’t think it helps at all. But if `Functor` is really “endofunctor in the category of Scal”, I’d expect `FunctorK` to mean the same thing as `K` in other instances, which basically adds “endofunctors” to the description – i.e., “endofunctor in the category of endofunctors in Scal”.

I’d also be happy with `Functor` being renamed to `Endofunctor` and having `EndofunctorK` and something like `LowerFunctor` (for a functor from the category of endofunctors to the category of Skal).

Also agreed that none of this should be in cats-core. I would like to find some time to work with `-Ykind-polymorphism` and to define these things as specializations of kind-polymorphic endofunctors, etc.

This sort of thing has been started in a few different libraries already – khats (which is too homophonically-named to be useful), Griffins (which never really got off the ground and mostly lives inside a Matryoshka branch), and I think one other that I have forgotten 😕